### PR TITLE
Fix #1135 Create persistent sessions for mapper and agent with non-zero QoS

### LINKED ISSUE #1135: Restart operation stuck in EXECUTING state
**Describe the bug**
After triggering a restart operation, the device was restarted successfully. However, the operation status in Cumulocity is still in EXECUTING.

**To Reproduce**
Trigger a restart operation from Cumulocity.

**Expected behavior**
The tedge-agent reports "SUCCESSFUL" to Cumulocity.

**Screenshots**
Screenshot from 2022-05-11 09-09-55

**Environment (please complete the following information):**
 - OS [Raspberry Pi OS 10]
 - Hardware [Rapsberry Pi 4 8GB]
 - System-Architecture [Linux raspberrypi 5.10.63-v7l+ #1496 SMP Wed Dec 1 15:58:56 GMT 2021 armv7l GNU/Linux]
 - thin-edge.io version [0.6.3 from ci-pipeline workflow, [here](https://github.com/thin-edge/thin-edge.io/actions/runs/2294734877)]

**Additional context**

### PR DESCRIPTION
**Pull request by albinsuresh**
## Proposed changes

Create persistent sessions for mapper and agent with non-zero QoS

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Improvement (general improvements like code refactoring that doesn't explicitly fix a bug or add any new functionality)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Paste Link to the issue
<br/>

## Checklist

- [ ] I have read the [CONTRIBUTING](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTING.md) doc
- [ ] I have signed the [CLA](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTOR-LICENSE-AGREEMENT.md) (in all commits with git commit -s)
- [ ] I ran `cargo fmt` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [ ] I used `cargo clippy` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

File: crates/common/mqtt_channel/src/session.rs [hunk 1]
```diff
@@ -26,10 +26,11 @@ pub async fn init_session(config: &Config) -> Result<(), MqttError> {
...
                 };
                 let subscriptions = config.subscriptions.filters();
-                if subscriptions.is_empty() {
-                    break;
+                for subscription in subscriptions {
+                    mqtt_client
+                        .subscribe(subscription.path, config.subscriptions.qos)
+                        .await?;
                 }
-                mqtt_client.subscribe_many(subscriptions).await?;
```
Comment (didier-wenzek) [inline crates/common/mqtt_channel/src/session.rs:32]: Same comments as for `Connection::open()` plus a really important difference.

Here, the case `_ => ()` actually silently discards any received messages, since there is no `Ok(Event::Incoming(Packet::Publish(msg)))` match case.

This means that the idea of creating an MQTT session with a `--init` option is fragile and even buggy if done after the very first installation. If called twice or once the mapper or the agent has already been launched, this command will consume all the pending messages preventing them to be processed.

=> We need to rethink the way MQTT daemons are started. They must not assume that all their peers are ready to consume messages and they must handle with extra care all init messages - as the initial software list request sent by the c8y mapper on start.

File: crates/core/tedge_agent/src/agent.rs [hunk 2]
```diff
@@ -649,7 +669,7 @@ mod tests {
 
     use super::*;
 
-    const SLASH_RUN_PATH_TEDGE_AGENT_RESTART: &str = "tedge_agent/tedge_agent_restart";
+    const SLASH_RUN_PATH_TEDGE_AGENT_RESTART: &str = "lock/tedge_agent_restart";
```
Comment (didier-wenzek) [inline crates/core/tedge_agent/src/agent.rs:672]: Why this change?

File: crates/common/mqtt_channel/src/connection.rs [hunk 3]
```diff
@@ -117,10 +117,11 @@ impl Connection {
...
                     };
                     let subscriptions = config.subscriptions.filters();
-                    if subscriptions.is_empty() {
-                        break;
+                    for subscription in subscriptions {
+                        mqtt_client
+                            .subscribe(subscription.path, config.subscriptions.qos)
+                            .await?;
                     }
-                    mqtt_client.subscribe_many(subscriptions).await?;
```
[the remark this answers is not in the document]
Comment (albinsuresh) [reply crates/common/mqtt_channel/src/connection.rs:123]: I misread that `subscribe_many` API. You're right and the test `mqtt_channel::tests::creating_a_session` even validates this behaviour. So, reverting this unnecessary change.

File: crates/common/mqtt_channel/src/session.rs (hunk 1 above)
Comment (albinsuresh) [reply crates/common/mqtt_channel/src/session.rs:32]: Will revert.

File: crates/core/tedge_agent/src/agent.rs [hunk 4]
```diff
@@ -542,6 +543,7 @@ impl SmAgent {
         responses: &mut impl PubChannel,
         topic: &Topic,
     ) -> Result<(), AgentError> {
+        debug!("Persisting restart operation");
```
Comment (PradeepKiruvale) [inline crates/core/tedge_agent/src/agent.rs:546]: remove debug statement.

File: crates/core/tedge_agent/src/agent.rs [hunk 5]
```diff
@@ -551,10 +553,14 @@ impl SmAgent {
         let () = responses
             .publish(Message::new(topic, executing_response.to_bytes()?))
             .await?;
+
+        debug!("Creating restart operation signal file");
```
Comment (PradeepKiruvale) [inline crates/core/tedge_agent/src/agent.rs:557]: remove debug statement.

File: crates/core/tedge_agent/src/agent.rs [hunk 6]
```diff
@@ -551,10 +553,14 @@ impl SmAgent {
...
         let () = restart_operation::create_slash_run_file(&self.config.run_dir)?;
 
+        debug!("Sync all pending writes to disk");
```
Comment (PradeepKiruvale) [inline crates/core/tedge_agent/src/agent.rs:560]: remove debug statement.

File: crates/core/tedge_agent/src/agent.rs [hunk 7]
```diff
@@ -551,10 +553,14 @@ impl SmAgent {
...
         let _process_result = std::process::Command::new("sudo").arg("sync").status();
-        // state = "Restarting"
+
+        debug!("Run restart command");
```
Comment (PradeepKiruvale) [inline crates/core/tedge_agent/src/agent.rs:563]: remove debug statement.

File: crates/core/tedge_mapper/src/core/mapper.rs [hunk 8]
```diff
@@ -130,6 +130,7 @@ impl Mapper {
     }
 
     async fn process_message(&mut self, message: Message) {
+        debug!("Received message: {:?}", message);
```
Comment (PradeepKiruvale) [inline crates/core/tedge_mapper/src/core/mapper.rs:133]: remove debug statement.

File: crates/core/tedge_agent/src/agent.rs [hunk 9]
```diff
@@ -619,10 +626,9 @@ impl SmAgent {
             };
 
             let response = SoftwareRequestResponse::new(&id, status);
+            let operation_status_message = Message::new(topic, response.to_bytes()?);
 
-            let () = responses
-                .publish(Message::new(topic, response.to_bytes()?))
-                .await?;
+            let () = responses.publish(operation_status_message).await?;
```
Comment (PradeepKiruvale) [inline crates/core/tedge_agent/src/agent.rs:631]: I see no reason for this change here.

File: crates/core/tedge_agent/src/agent.rs (hunk 4 above)
Comment (albinsuresh) [reply crates/core/tedge_agent/src/agent.rs:546]: Why? It's a debug log and not `info` or `warn`. It can help with debugging in future.

**Comment by albinsuresh** [comment]
Closing this PR and will be replaced with the fix proposed in the last comment.
